### PR TITLE
[Issue #3871] File activation for unpackaged App would fail when file path contains unicode characters or special characters

### PR DESCRIPTION
**Pull request by MuyuanMS**
## Background
For unpackaged App, when opening a file with type registered with [RegisterForFileTypeActivation()](https://learn.microsoft.com/windows/windows-app-sdk/api/winrt/microsoft.windows.applifecycle.activationregistrationmanager.registerforfiletypeactivation?view=windows-app-sdk-1.6), current design is to execute the app with encoded protocol as argument (`----ms-protocol:ms-encodedlaunch:App?ContractId=Windows.File&Verb=open&File=%1`)

E.g.
`"C:\MyApp.exe" "----ms-protocol:ms-encodedlaunch:App?ContractId=Windows.File&Verb=open&File=C:\myfile.foo"`

## Expectation
The expectation is when the app calls `GetActivatedEventArgs()`, the returned value should has a type of `ExtendedActivationKind.File` with file path/name in the data.

## Problem
However, when the file path contains unicode characters or special characters like `&`, the file activation would fail as the encoded protocol is not correctly parsed. E.g.
### With unicode characters:
`"C:\MyApp.exe" "----ms-protocol:ms-encodedlaunch:App?ContractId=Windows.File&Verb=open&File=C:\你好.foo"`
This would result in activation type being parsed as `ExtendedActivationKind.Protocol`

### With special character:
`"C:\MyApp.exe" "----ms-protocol:ms-encodedlaunch:App?ContractId=Windows.File&Verb=open&File=C:\a&b.foo"`
This would crash the app on calling `GetActivatedEventArgs()`

## Root cause
The query in `winrt::Windows::Foundation::Uri` is parsed with `QueryParsed()`, which would split on `&` and cause problem when having unicode characters.

## Solution
The best way to address the issue, is to change the activation protocol design, such that file path, which may contain unicode characters or special characters, is not placed in the protocol uri not designed to handle file paths. 
E.g. instead of `----ms-protocol:ms-encodedlaunch:App?ContractId=Windows.File&Verb=open&File=%1`, we put the file path as separate parameter like:
`"----ms-protocol:ms-encodedlaunch:App?ContractId=Windows.File&Verb=open" "%1"`

However, this would require design changes in multiple components including registering, serializing and deserializing protocols. For now, this PR is a simple fix just on parsing logic.

Instead of using Uri.QueryParsed(), we use simple string operations to parse the uri query. (E.g. substring from "&File=" to get the file path). 

## Test
During the fix I locally created a sample app launched with below methods to test the change:

1. activation from file `abc.txt`
2. activation from file `你好.txt`
3. activation from file `a&b.txt`
4. activation from file `😀.txt`

Previously, 2 would give protocol type, and 3 would crash the app.
After fix, all three cases give file type, with correct file path contained in the returned value.

I have also added a functional test activating unpackaged app via a file with unicode characters and special character `&`

# Auto Generated PR Description

A microsoft employee must use /azp run to validate using the pipelines below.

WARNING:
Comments made by azure-pipelines bot maybe inaccurate.
Please see pipeline link to verify that the build is being ran.

For status checks on the main branch, please use TransportPackage-Foundation-PR
(https://microsoft.visualstudio.com/ProjectReunion/_build?definitionId=81063&_a=summary)
and run the build against your PR branch with the default parameters.

### CODE REVIEW

File: dev/AppLifecycle/FileActivatedEventArgs.h [hunk 1]
```diff
@@ -47,6 +47,24 @@ namespace winrt::Microsoft::Windows::AppLifecycle::implementation
...
         {
+            // Standard serialize has a query of the form: "?ContractId=Windows.File&Verb=<verb>&File=<file_path>"
+            // If the query has the exact same format, then we can deserialize it with simple string operations,
+            // as QueryParsed() can cause issues when file path contains unicode characters or specical characters like &.
+            std::wstring serializedQueryHeader = L"?" + std::wstring(c_contractIdKeyName) + L"=" + std::wstring(c_fileContractId) + L"&Verb=";
+            int headerLen = (int)wcslen(serializedQueryHeader.c_str());
+            if (CompareStringOrdinal(uri.Query().c_str(), headerLen, serializedQueryHeader.c_str(), -1, TRUE) == CSTR_EQUAL)
+            {
+                int queryLen = (int)wcslen(uri.Query().c_str());
+                const std::wstring query = uri.Query().c_str();
```
Comment (Scottj1s) [inline dev/AppLifecycle/FileActivatedEventArgs.h:58]: consider hoisting this up to reuse on lines 55, 57

File: dev/AppLifecycle/FileActivatedEventArgs.h [hunk 2]
```diff
@@ -47,6 +47,24 @@ namespace winrt::Microsoft::Windows::AppLifecycle::implementation
...
+                int verbEnd = query.find(L"&", headerLen);
```
Comment (Scottj1s) [inline dev/AppLifecycle/FileActivatedEventArgs.h:59]: need error checking for [std::string::npos](https://en.cppreference.com/w/cpp/string/basic_string/npos)

File: dev/AppLifecycle/FileActivatedEventArgs.h [hunk 3]
```diff
@@ -47,6 +47,24 @@ namespace winrt::Microsoft::Windows::AppLifecycle::implementation
...
+                auto verb = winrt::to_hstring(query.substr(headerLen, verbEnd - headerLen).c_str());
+
+                int filePos = query.find(L"&File=");
```
Comment (Scottj1s) [inline dev/AppLifecycle/FileActivatedEventArgs.h:62]: need error checking for [std::string::npos](https://en.cppreference.com/w/cpp/string/basic_string/npos)

File: dev/AppLifecycle/FileActivatedEventArgs.h (hunk 2 above)
Comment (Scottj1s) [reply dev/AppLifecycle/FileActivatedEventArgs.h:59]: also, isn't this going to find the first '&' in the query string, prior to Verb?  it might be better to search remaining substrings as you go - strip the query header off to create a std::wstring verb = "Verb=<verb>&File=<file_path>", and strip that off to create a std::wstring file = "File=<file_path>"

File: dev/AppLifecycle/ExtensionContract.h [hunk 4]
```diff
@@ -37,6 +37,17 @@ namespace winrt::Microsoft::Windows::AppLifecycle::implementation
 
     inline std::tuple<ExtendedActivationKind, winrt::Windows::Foundation::IInspectable> DecodeActivatedEventArgs(winrt::Windows::Foundation::Uri const& uri)
     {
+        // If the Uri strictly follows uri format from GenerateEncodedLaunchUri, first query parameter is ContractId
```
Comment (dhoehna) [inline dev/AppLifecycle/ExtensionContract.h:40]: What if this is not true?  If the Uri does not strictly follow the uri format from GenerateEncodedLaunchUri?

File: dev/AppLifecycle/ExtensionContract.h [hunk 5]
```diff
@@ -37,6 +37,17 @@ namespace winrt::Microsoft::Windows::AppLifecycle::implementation
...
+        // This is to avoid calling uri.QueryParsed() as it can cause issues when uri contains unicode characters or special characters like &.
+        for (const auto& extension : c_extensionMap)
+        {
+            std::wstring contractIdQuery = L"?" + std::wstring(c_contractIdKeyName) + L"=" + std::wstring(extension.contractId);
+            if (CompareStringOrdinal(uri.Query().c_str(), (int)wcslen(contractIdQuery.c_str()), contractIdQuery.c_str(), -1, TRUE) == CSTR_EQUAL)
```
Comment (dhoehna) [inline dev/AppLifecycle/ExtensionContract.h:45]: `CompareStringOrdinal` looks for binary equality.  This looks like a linguistic comparison and should use the appropriate method.  Please use [comparestringex](https://learn.microsoft.com/windows/win32/api/stringapiset/nf-stringapiset-comparestringex)

Comment (dhoehna) [inline dev/AppLifecycle/ExtensionContract.h:45]: For the 2nd parameter, why the length of `contractIdQuery` and not `uri`?

File: dev/AppLifecycle/ExtensionContract.h [hunk 6]
```diff
@@ -37,6 +37,17 @@ namespace winrt::Microsoft::Windows::AppLifecycle::implementation
 
     inline std::tuple<ExtendedActivationKind, winrt::Windows::Foundation::IInspectable> DecodeActivatedEventArgs(winrt::Windows::Foundation::Uri const& uri)
     {
+        // If the Uri strictly follows uri format from GenerateEncodedLaunchUri, first query parameter is ContractId
+        // This is to avoid calling uri.QueryParsed() as it can cause issues when uri contains unicode characters or special characters like &.
+        for (const auto& extension : c_extensionMap)
+        {
+            std::wstring contractIdQuery = L"?" + std::wstring(c_contractIdKeyName) + L"=" + std::wstring(extension.contractId);
```
Comment (dhoehna) [inline dev/AppLifecycle/ExtensionContract.h:44]: I'm having a hard time understanding what this method does.  The difference between this loop, and the loop below is comparing the string with both the contract name and contract id.  Further, it uses the same methods to compare the strings.

File: dev/AppLifecycle/ExtensionContract.h [hunk 7]
```diff
@@ -37,6 +37,17 @@ namespace winrt::Microsoft::Windows::AppLifecycle::implementation
...
+            if (CompareStringOrdinal(uri.Query().c_str(), (int)wcslen(contractIdQuery.c_str()), contractIdQuery.c_str(), -1, TRUE) == CSTR_EQUAL)
+            {
+                return { extension.kind, extension.factory(uri) };
+            }
+        }
+
```
Comment (dhoehna) [inline dev/AppLifecycle/ExtensionContract.h:50]: What is true if code reaches here?

File: dev/AppLifecycle/FileActivatedEventArgs.h [hunk 8]
```diff
@@ -47,6 +47,24 @@ namespace winrt::Microsoft::Windows::AppLifecycle::implementation
 
         static winrt::Windows::Foundation::IInspectable Deserialize(winrt::Windows::Foundation::Uri const& uri)
         {
+            // Standard serialize has a query of the form: "?ContractId=Windows.File&Verb=<verb>&File=<file_path>"
+            // If the query has the exact same format, then we can deserialize it with simple string operations,
```
Comment (dhoehna) [inline dev/AppLifecycle/FileActivatedEventArgs.h:51]: What will happen if the query is not the same format?

Comment (dhoehna) [inline dev/AppLifecycle/FileActivatedEventArgs.h:51]: NIT: Don't use `we` in code comments.

File: dev/AppLifecycle/FileActivatedEventArgs.h (hunk 1 above)
Comment (dhoehna) [reply dev/AppLifecycle/FileActivatedEventArgs.h:58]: +1 considering the duplicate string code is in a different file.

File: dev/AppLifecycle/FileActivatedEventArgs.h [hunk 9]
```diff
@@ -47,6 +47,24 @@ namespace winrt::Microsoft::Windows::AppLifecycle::implementation
...
+            // as QueryParsed() can cause issues when file path contains unicode characters or specical characters like &.
+            std::wstring serializedQueryHeader = L"?" + std::wstring(c_contractIdKeyName) + L"=" + std::wstring(c_fileContractId) + L"&Verb=";
+            int headerLen = (int)wcslen(serializedQueryHeader.c_str());
```
Comment (dhoehna) [inline dev/AppLifecycle/FileActivatedEventArgs.h:54]: auto.

File: dev/AppLifecycle/FileActivatedEventArgs.h (hunk 3 above)
Comment (dhoehna) [inline dev/AppLifecycle/FileActivatedEventArgs.h:62]: is `.find()` case-sensitive?

**Comment by dhoehna** [comment]
Please add tests.

File: dev/AppLifecycle/ExtensionContract.h (hunk 5 above)
Comment (MuyuanMS) [reply dev/AppLifecycle/ExtensionContract.h]: Uri parsing and checking for exact match sounds more like ordinal comparison, like existing code for parsing contract id - Or are there cases where protocols to be put in registry could be localized?

File: dev/AppLifecycle/ExtensionContract.h (hunk 6 above)
Comment (MuyuanMS) [reply dev/AppLifecycle/ExtensionContract.h]: Sorry for the poor readability in logic - I have refactored it a little bit.
Point is when dealing with encoded protocols for files, when file contains special characters, `QueryParsed()` function would not parse correctly (it would return empty if unicode characters are involved) and the current loop may not run at all.

File: dev/AppLifecycle/FileActivatedEventArgs.h (hunk 8 above)
Comment (MuyuanMS) [reply dev/AppLifecycle/FileActivatedEventArgs.h]: This encoded launch is for WinAppSDK apps only - which means all registration should be from the code here. There's no public documentation suggesting this encoded protocol can be customized.
Without the assumption, there's no way to guarantee correct parsing, as file path could contain special characters like `&`
Never the less, I have refactored a little bit to loosen the assumption to be the query should contain 3 and only 3 pairs, for ContractId, Verb, and File.

Comment (MuyuanMS) [reply dev/AppLifecycle/FileActivatedEventArgs.h]: Got it, thanks!

File: dev/AppLifecycle/FileActivatedEventArgs.h (hunk 2 above)
Comment (MuyuanMS) [reply dev/AppLifecycle/FileActivatedEventArgs.h]: The second argument in `find()` indicates the index to start the search, so it finds the first `&` after `Verb=`

File: dev/AppLifecycle/FileActivatedEventArgs.h (hunk 3 above)
Comment (MuyuanMS) [reply dev/AppLifecycle/FileActivatedEventArgs.h]: Yes, and the existing logic also treats protocol query as case sensitive

File: dev/AppLifecycle/ExtensionContract.h (hunk 4 above)
Comment (dhoehna) [reply dev/AppLifecycle/ExtensionContract.h]: I see.  I would suggest to add the reasoning why the assumption is safe to make.

File: dev/AppLifecycle/ExtensionContract.h (hunk 5 above)
Comment (dhoehna) [reply dev/AppLifecycle/ExtensionContract.h]: I see URI and I think of a user making the URI, or a part of it.  Language comparison.  Further, unless I am incorrect, UNICODE characters can be represented with a different number of bytes.  

If the uri is generated by code, I believe a binary comparison would suffice.

File: dev/AppLifecycle/ExtensionContract.h (hunk 6 above)
Comment (dhoehna) [reply dev/AppLifecycle/ExtensionContract.h]: Coolio.

File: dev/AppLifecycle/FileActivatedEventArgs.h (hunk 8 above)
Comment (dhoehna) [reply dev/AppLifecycle/FileActivatedEventArgs.h]: I apologize.  I should be more clear in what I'm asking for.  Can you add an assert, or some other code, that checks to make sure the queries have the exact same format?

I'm only thinking ahead by 1-5 years when this code breaks because the assumption is not true anymore.

File: dev/AppLifecycle/ExtensionContract.h [hunk 10]
```diff
@@ -52,6 +52,14 @@ namespace winrt::Microsoft::Windows::AppLifecycle::implementation
             }
         }
 
+        // QueryParsed() function would return empty when it's a file contract with unicode characters in file path
+        // Thus following additional check for file contract is needed
+        auto fileContractUri = GenerateEncodedLaunchUri(L"App", c_fileContractId);
```
Comment (dhoehna) [inline dev/AppLifecycle/ExtensionContract.h:57]: NIT: I can't comment on the code above here.  I'll leave the comment here instead.  Line 40-53.  Prefer to check for the incorrect states and `continue` early instead of having nested scopes.

**Comment by dhoehna** [comment]
The code looks good to me.  Please add automated tests.

File: dev/AppLifecycle/FileActivatedEventArgs.h [hunk 11]
```diff
@@ -47,9 +47,46 @@ namespace winrt::Microsoft::Windows::AppLifecycle::implementation
 
         static winrt::Windows::Foundation::IInspectable Deserialize(winrt::Windows::Foundation::Uri const& uri)
         {
-            auto query = uri.QueryParsed();
-            auto verb = query.GetFirstValueByName(L"Verb");
-            auto file = query.GetFirstValueByName(L"File");
+            auto parsedQuery = uri.QueryParsed();
+            if (parsedQuery.Size() == 3)
```
Comment (dhoehna) [inline dev/AppLifecycle/FileActivatedEventArgs.h:51]: Why 3?  Is this a safe upper limit, or is 3 the magic number of queries with UNICODE characters and special characters?

Comment (dhoehna) [reply dev/AppLifecycle/FileActivatedEventArgs.h:51]: I just now read the comment.  I suggest to move it up to explain the number 3.

File: dev/AppLifecycle/ExtensionContract.h (hunk 10 above)
Comment (MuyuanMS) [reply dev/AppLifecycle/ExtensionContract.h:57]: The first `if` statement on Line 42 does the job, if I'm understanding this comment correctly. It will only enter nested loops if the query name is `ContractId`

File: dev/AppLifecycle/FileActivatedEventArgs.h (hunk 11 above)
Comment (MuyuanMS) [reply dev/AppLifecycle/FileActivatedEventArgs.h:54]: Done

File: dev/AppLifecycle/FileActivatedEventArgs.h [hunk 12]
```diff
@@ -47,9 +47,46 @@ namespace winrt::Microsoft::Windows::AppLifecycle::implementation
...
+                return make<FileActivatedEventArgs>(verb, file);
+            }
+
+            const std::wstring query = uri.Query().c_str();
+            auto queryLength = query.length();
+
+            auto verbBegin = query.find(L"&Verb=");
+            if (verbBegin == std::wstring::npos)
+            {
+                throw winrt::hresult_invalid_argument();
```
Comment (dhoehna) [inline dev/AppLifecycle/FileActivatedEventArgs.h:67]: for this, and all other `winrt::hresult_invalid_argument` please add a message indicating the incorrect argument and why it is incorrect.

File: test/AppLifecycle/FunctionalTests.cpp [hunk 13]
```diff
@@ -128,6 +131,42 @@ namespace Test::AppLifecycle
...
+
+            // Cleanup any leftover data from previous runs i.e. ensure we running with a clean slate
+            try
+            {
+                Execute(L"AppLifecycleTestApp.exe", L"/UnregisterFile", g_deploymentDir);
+                WaitForEvent(event, m_failed);
+            }
+            catch (...)
+            {
+                //TODO:Unregister should not fail if ERROR_FILE_NOT_FOUND | ERROR_PATH_NOT_FOUND
```
Comment (dhoehna) [inline test/AppLifecycle/FunctionalTests.cpp:147]: Instead of a comment, add the check here.  That way, if Unregister does fail if ERROR_FILE_NOT_FOUND or ERROR_PATH_NOT_FOUND developers will be notified.

**Comment by dhoehna** [comment]
Thanks for adding the test. :)

**Review by dhoehna — APPROVED** [review]

File: test/AppLifecycle/FunctionalTests.cpp (hunk 13 above)
Comment (MuyuanMS) [reply test/AppLifecycle/FunctionalTests.cpp:147]: This is copied from other testcases unregistering protocol or files - If this would have failed due to file not found or path not found, the following execute statements would fail as well anyway and developer can indeed know it.

File: dev/AppLifecycle/FileActivatedEventArgs.h (hunk 12 above)
Comment (MuyuanMS) [reply dev/AppLifecycle/FileActivatedEventArgs.h]: Added

**Review by dhoehna — APPROVED** [review]
